core/domain/story_domain.py

```diff
@@ -234,6 +234,37 @@ def require_valid_node_id(cls, node_id):
             raise utils.ValidationError(
                 'Invalid node_id: %s' % node_id)
 
+    @classmethod
+    def require_valid_thumbnail_filename(cls, thumbnail_filename):
+        """Checks whether the thumbnail filename of the node is a valid
+            one.
+
+        Args:
+            thumbnail_filename: str. The thumbnail filename to validate.
+        """
+        if thumbnail_filename is not None:
+            if not isinstance(thumbnail_filename, python_utils.BASESTRING):
+                raise utils.ValidationError(
+                    'Expected thumbnail filename to be a string, received %s'
+                    % thumbnail_filename)
+            if thumbnail_filename.rfind('.') == 0:
+                raise utils.ValidationError(
+                    'Thumbnail filename should not start with a dot.')
+            if '/' in thumbnail_filename or '..' in thumbnail_filename:
+                raise utils.ValidationError(
+                    'Thumbnail filename should not include slashes or '
+                    'consecutive dot characters.')
+            if '.' not in thumbnail_filename:
+                raise utils.ValidationError(
+                    'Thumbnail filename with no extension.')
+
+            dot_index = thumbnail_filename.rfind('.')
+            extension = thumbnail_filename[dot_index + 1:].lower()
+            if extension != 'png':
```

Comment:
Do you think it is OK to accept svgs? I think we might actually want to prefer those (and actually forbid png) because they are better for display on different viewport sizes.

Reply:
Currently, we don't allow upload of SVG in the UI. I'll investigate this. Thanks!

Reply:
The output of Cropper (third party library we are using for cropping) returns a canvas. Need to figure out a way to convert canvas to SVG -- quick search tells me we have to introduce another library for this :P.

Reply:
Introducing another library sounds reasonable if necessary.

Reply:
Done. But I'm allowing upload of any format in the UI but then convert it to SVG and send it over to the backend to save.

Reply:
Hm. Does that actually work? I.e. can a PNG be converted to SVG "losslessly"?

/cc @BenHenning since he may have thoughts on this.

Reply:
I think what is currently happening is that it just creates an SVG document and adds an image tag inside with a png base64 href. 

I took a look and this is what the SVG code looks like:

```
"<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="120" height="67"><defs/><g><image width="120" height="67" preserveAspectRatio="none" xlink:href="data:image/png;base64,qwerty="/><g transform="translate(0,0)"/></g></svg>"
```

Reply:
That doesn't seem ideal to me. I agree that should prefer SVGs, but I'm a bit hesitant to outright ban them. I think we still need to retain support for bitmaps, but we'll need to upload different resolutions of them (e.g. https://docs.google.com/document/d/1J0Mvs1pe6ZQdHKAWgChb4AEXdqB_DovQevf-iXbu0Bc/edit#heading=h.9kdcd9zhv9rx). Unfortunately, I haven't had time to clarify what is meant by different sizes of images beyond what's in that doc. Ideally, we would require all images be in one of two formats (without any conversion between them):
1. SVGs (we upload these as-is with the user-recommended size or whatever size is specified within the SVG)

2. Webp files (we could convert PNG/JPG clientside to Webp for convenience), where the user has to specify one image for each display density it might show up on. I envision an editor screen where the user uploads an initial image and we should them how it will look shrunk/expanded based on the target display densities, and they can choose to upload additional images for each density to "fix" it. This would be a reusable component for all image upload functionality in the frontend, and could technically show the SVG in different display densities as well if we care.

I'm not quite sure how that affects the work being done in this domain model, but does it help clarify things or further muddle them? :)

Reply:
Thanks @BenHenning for clarifying things!

I have just one more question: In the case of SVG upload, I was wondering if there is a way to crop the image using svg attributes such as viewBox / height / width alone? I'm currently using this [library](https://fengyuanchen.github.io/cropperjs/) to crop images but it seems to internally convert SVG to PNG. However, it does provide data such as the crop dimensions and left, top positions for the cropped image (click on 'getData', 'getCropboxData' in the [link](https://fengyuanchen.github.io/cropperjs/)). I was wondering if we can crop an SVG image using this library in the UI but internally we just modify the original SVG by simply changing the SVG attributes according to the cropped image data from the library. Is this possible?

Reply:
We want to keep the underlying format as SVG that we store in the backend, so the cropping would need to be a client-side solution (e.g. we pass the SVG & cropping parameters to the frontends to decide how to crop). Oppia web can sort of do whatever it wants--either convert to PNG & crop as this library does, or do another solution. I don't have a strong preference there, as long as we are keeping the SVG file in the backend so that we can natively render it on Android (we'll also need to solve the cropping problem, and without being able to use the reference library).

Does that help?

Reply:
Cool, sounds good. I'll investigate if it is possible to crop the SVG in the frontend in someother way. Thanks!

Just to check, I'm able to convert the cropped canvas image that the library provides into an SVG like in my previous comment ([here])(https://github.com/oppia/oppia/pull/8498#discussion_r380517226)). Do you think this sort of conversion is fine atleast for SVG uploads? For PNG and JPG uploads we can convert to WebP directly from canvas (eg. using canvas.toDataURL('image/webp')) before uploading it to the backend.

Reply:
Ah sorry I wasn't clear earlier--I'm not keen on wrapping bitmap images in an SVG container. SVGs ought to truly be scalable, and bitmaps should have multiple "mip" levels (as in [mipmap](https://en.wikipedia.org/wiki/Mipmap)). We should make this distinction everywhere: the backend should have two ways to store images, SVG and bitmap, where bitmaps have potentially multiple images associated with them (always webp format, and each image is specific to a certain resolution). Frontends then can encounter either image type and will need to do their best to render them correctly (e.g. by picking the right mip level, or scaling the SVG).

Does that help clarify?